src/lib/hooks/swap/useSwapCallback.tsx

```diff
@@ -0,0 +1,175 @@
+import { BigNumber } from '@ethersproject/bignumber'
+// eslint-disable-next-line no-restricted-imports
+import { t, Trans } from '@lingui/macro'
+import { Trade } from '@uniswap/router-sdk'
+import { Currency, Percent, TradeType } from '@uniswap/sdk-core'
+import { Trade as V2Trade } from '@uniswap/v2-sdk'
+import { Trade as V3Trade } from '@uniswap/v3-sdk'
+import { DEFAULT_DEADLINE_FROM_NOW } from 'constants/misc'
+import useActiveWeb3React from 'hooks/useActiveWeb3React'
+import useCurrentBlockTimestamp from 'hooks/useCurrentBlockTimestamp'
+import useENS from 'hooks/useENS'
+import { SignatureData } from 'hooks/useERC20Permit'
+import { useSwapCallArguments } from 'hooks/useSwapCallArguments'
+import { ReactNode, useMemo } from 'react'
+import { calculateGasMargin } from 'utils/calculateGasMargin'
+import isZero from 'utils/isZero'
+import { swapErrorToUserReadableMessage } from 'utils/swapErrorToUserReadableMessage'
+
+type AnyTrade =
+  | V2Trade<Currency, Currency, TradeType>
+  | V3Trade<Currency, Currency, TradeType>
+  | Trade<Currency, Currency, TradeType>
+
+enum SwapCallbackState {
+  INVALID,
+  LOADING,
+  VALID,
+}
+
+interface SwapCall {
+  address: string
+  calldata: string
+  value: string
+}
+
+interface SwapCallEstimate {
+  call: SwapCall
+}
+
+interface SuccessfulCall extends SwapCallEstimate {
+  call: SwapCall
+  gasEstimate: BigNumber
+}
+
+interface FailedCall extends SwapCallEstimate {
+  call: SwapCall
+  error: Error
+}
+
+// returns a function that will execute a swap, if the parameters are all valid
+// and the user has approved the slippage adjusted input amount for the trade
+export function useSwapCallback(
```

Comment:
I would prefer that `lib/hooks/swap/useSwapCallback` be a generic swap callback, and that then be wrapped in `hooks/useSwapCallback`, similar to how transactions does it. This avoids coupling the swap and transactions state, and avoids having the code duplicated.